cmd/registry/patch/artifact.go

```diff
@@ -43,46 +48,99 @@ func ExportArtifact(ctx context.Context, client *gapic.RegistryClient, message *
 		}
 		message.Contents = body.Data
 	}
-	var artifact Artifact
+	artifact, err := newArtifact(message)
+	if err != nil {
+		return nil, nil, err
+	}
+	var b bytes.Buffer
+	err = yamlEncoder(&b).Encode(artifact)
+	if err != nil {
+		return nil, nil, err
+	}
+	return b.Bytes(), &artifact.Header, nil
+}
+
+func newArtifact(message *rpc.Artifact) (*Artifact, error) {
+	var artifact *Artifact
 	var err error
 	switch message.GetMimeType() {
+	case DisplaySettingsMimeType:
+		artifact, err = newDisplaySettings(message)
 	case LifecycleMimeType:
 		artifact, err = newLifecycle(message)
 	case ManifestMimeType:
 		artifact, err = newManifest(message)
+	case ReferenceListMimeType:
+		artifact, err = newReferenceList(message)
 	case TaxonomyListMimeType:
 		artifact, err = newTaxonomyList(message)
 	default:
 		artifact, err = newUnknownArtifact(message)
 	}
-	if err != nil {
-		return nil, nil, err
+	return artifact, err
+}
+
+func (a *Artifact) UnmarshalYAML(node *yaml.Node) error {
+	// https://stackoverflow.com/questions/66709979/dynamically-parse-yaml-field-to-one-of-a-finite-set-of-structs-in-go
+	type Alias Artifact
+	type Wrapper struct {
+		*Alias `yaml:",inline"`
+		Data   yaml.Node `yaml:"data"`
 	}
-	var b bytes.Buffer
-	err = yamlEncoder(&b).Encode(artifact)
+	wrapper := &Wrapper{Alias: (*Alias)(a)}
+	if err := node.Decode(wrapper); err != nil {
+		return err
+	}
+	switch a.Kind {
+	case "DisplaySettings":
+		a.Data = new(DisplaySettingsData)
+	case "Lifecycle":
+		a.Data = new(LifecycleData)
+	case "Manifest":
+		a.Data = new(ManifestData)
+	case "ReferenceList":
+		a.Data = new(ReferenceListData)
+	case "TaxonomyList":
+		a.Data = new(TaxonomyListData)
+	default:
+		return fmt.Errorf("unable to unmarshal %s", a.Kind)
+	}
+	return wrapper.Data.Decode(a.Data)
+}
+
+func (a *Artifact) MarshalYAML() (interface{}, error) {
+	type exportable struct {
+		Header `yaml:",inline"`
+		Data   ArtifactData `yaml:"data"`
+	}
+	return &exportable{
+		Header: a.Header,
+		Data:   a.Data,
+	}, nil
+}
```

Comment:
What is this?

Reply:
I found that this was necessary to properly export artifact YAMLs. You'll see that in the struct tags above (on line 31), the tag for "Data" is a dash ("-"). This prevents the importer from directly importing this value -- which it can't do because this value could be a struct of multiple types. See the StackOverflow link on line 84 for more on that.

Now, with "-" in the struct tag for Data, YAML exports will skip that field. Unlike imports, YAML exports have no problems with polymorphic fields, so we can get the field to export by setting its struct tag back to "data". Rather than mess with reflection or whatever is needed to tweak the struct tag, it seemed better to just copy the artifact into a new structure with appropriate tags and then export it.

I'm happy to do this in a better way if there is one, but I think having polymorphic Data inside artifacts will be a requirement (it's a Kubernetes resource pattern that I think someone was trying to follow when they asked that question on SO). I think in a future revision, we might want to have this code "float" up into a general reader of wrapped types that reads all of the YAMLs (APIs, Artifacts, etc) in one place in the code.

Reply:
I think this is worth explaining in a code comment. It's not obvious that we're implementing an interface that the `yaml` package is using at marshal time.

Reply:
OK, done!